libraries/SITL/SIM_Frame.cpp

```diff
@@ -278,16 +281,173 @@ static Frame supported_frames[] =
+        AP_HAL::panic("%s failed to load\n", model_json);
+    }
+    AP::FS().close(fd);
+
+    char *start = strchr(buf, '{');
+    if (!start) {
+        AP_HAL::panic("Invalid json %s", model_json);
+    }
+    free(fname);
+
     /*
-       scaling from total motor power to Newtons. Allows the copter
-       to hover against gravity when each motor is at hover_throttle
-    */
-    thrust_scale = (_mass * GRAVITY_MSS) / (num_motors * _hover_throttle);
+      remove comments, as not allowed by the parser
+     */
+    for (char *p = strchr(start,'#'); p; p=strchr(p+1, '#')) {
+        // clear to end of line
+        do {
+            *p++ = ' ';
+        } while (*p != '\n' && *p != '\r' && *p);
+    }
+
+    picojson::value obj;
+    std::string err = picojson::parse(obj, start);
+    if (!err.empty()) {
+        AP_HAL::panic("Failed to load %s: %s", model_json, err.c_str());
+        exit(1);
+    }
+
+    struct {
+        const char *label;
+        float &v;
+    } vars[] = {
+#define FRAME_VAR(s) { #s, model.s }
+        FRAME_VAR(mass),
+        FRAME_VAR(diagonal_size),
+        FRAME_VAR(refSpd),
+        FRAME_VAR(refAngle),
+        FRAME_VAR(refVoltage),
+        FRAME_VAR(refCurrent),
+        FRAME_VAR(refAlt),
+        FRAME_VAR(refTempC),
+        FRAME_VAR(maxVoltage),
+        FRAME_VAR(battCapacityAh),
+        FRAME_VAR(refBatRes),
+        FRAME_VAR(propExpo),
+        FRAME_VAR(refRotRate),
+        FRAME_VAR(hoverThrOut),
+        FRAME_VAR(pwmMin),
+        FRAME_VAR(pwmMax),
+        FRAME_VAR(spin_min),
+        FRAME_VAR(spin_max),
+        FRAME_VAR(slew_max),
+    };
+    static_assert(sizeof(model) == sizeof(float)*ARRAY_SIZE(vars), "incorrect model vars");
+
+    for (uint8_t i=0; i<ARRAY_SIZE(vars); i++) {
+        auto v = obj.get(vars[i].label);
+        if (v.is<picojson::null>()) {
+            // use default value
+            continue;
+        }
+        if (!v.is<double>()) {
+            AP_HAL::panic("Bad json type for %s: %s", vars[i].label, v.to_str().c_str());
+        }
+        vars[i].v = v.get<double>();
+    }
+
+    ::printf("Loaded model params from %s\n", model_json);
+}
+
+/*
+  initialise the frame
+ */
+void Frame::init(const char *frame_str, Battery *_battery)
+{
+    model = default_model;
+    battery = _battery;
+
+    const char *colon = strchr(frame_str, ':');
+    size_t slen = strlen(frame_str);
+    if (colon != nullptr && slen > 5 && strcmp(&frame_str[slen-5], ".json") == 0) {
+        load_frame_params(colon+1);
+    }
+    mass = model.mass;
+
+    const float drag_force = model.mass * GRAVITY_MSS * tanf(radians(model.refAngle));
+
+    float ref_air_density = get_air_density(model.refAlt);
+
+    areaCd = drag_force / (0.5 * ref_air_density * sq(model.refSpd));
+
+    terminal_rotation_rate = model.refRotRate;
+
+    float hover_thrust = mass * GRAVITY_MSS;
+    float hover_power = model.refCurrent * model.refVoltage;
```

Comment:
I'm abit late to the party here, Sorry.

This is not the hover power, its the flying along for the drag reference power. I guess you could add a hover voltage and current to the definition. Or you could add flying at a constant speed throttle and use that as the reference to get velocity max. This is assuming 100% battery to propeller efficiency in any-case so it might come out in the wash.

Reply:
In fact I think your sort of inferring that the `battery -> esc -> propeller` efficiency is `cos(refAngle)`. So the pitch angle you happen to use probably matters quite allot.

Possibly you could resolve if if you did a couple of different pitch angles for the steady state test.